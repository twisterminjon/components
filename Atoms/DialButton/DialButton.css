.dialbutton {
  display: flex;
  justify-content: center;
  align-items: center;

  border-radius: 100%;
  border: none;

  outline: none;

  width: 39px;
  height: 39px;
}

.dialbutton:hover {
  opacity: 0.9;
}

.dialbutton:focus {
  opacity: 0.9;
}

.dialbutton:active {
  transform: scale(0.95);
}

.dialbutton--enabled {
  color: var(--black);
  background-color: var(--statusbubble_available);
}

.dialbutton--disabled {
  background-color: transparent;
}

.dialbutton--disabled:hover {
  opacity: 1;
}

.dialbutton--dimmed {
  opacity: 0.5;
}
