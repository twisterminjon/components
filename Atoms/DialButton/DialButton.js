import React, { useContext } from 'react'
import PropTypes from 'prop-types'

import IconAnswerCall from '../../Atoms/Icons/IconAnswerCall'
import { USER_STATUS_LIST, USER_STATUS_AVAILABLE, USER_STATUS_BUSY, USER_STATUS_OFFLINE } from '../../../constants'
import { getTestId } from '@shared/helpers'
import { CurrentUserContext } from '@shared/providers'

import './DialButton.css'

DialButton.propTypes = {
  /** Adjusts the color based on a users status */
  status: PropTypes.oneOf(USER_STATUS_LIST),
  /** Function called when the button is clicked */
  onClick: PropTypes.func.isRequired,
  /** If true, will display in a dimmed state */
  dimmed: PropTypes.bool,
  /** ClassName for the wrapper */
  className: PropTypes.string,
  /** Style for the wrapper */
  style: PropTypes.object,
}

export default function DialButton({
  onClick,
  status = USER_STATUS_AVAILABLE,
  dimmed = false,
  className = '',
  style = {},
  ...rest
}) {
  const currentUser = useContext(CurrentUserContext)

  let disabled = status === USER_STATUS_BUSY || status === USER_STATUS_OFFLINE
  if (!disabled) disabled = !currentUser.micCamAvailable

  const disabledClass = disabled ? 'dialbutton--disabled' : 'dialbutton--enabled'
  const dataTestId = getTestId(rest, 'button-start-call')

  return (
    <button
      aria-label="dial"
      disabled={disabled}
      className={`dialbutton ${disabledClass} ${dimmed ? 'dialbutton--dimmed' : ''} ${className}`.trim()}
      style={style}
      onClick={onClick}
      data-testid={dataTestId}>
      <IconAnswerCall
        style={{ transform: 'scale(0.7)' }}
        color={disabled ? 'var(--statusbubble_offline)' : 'var(--black)'}
      />
    </button>
  )
}
