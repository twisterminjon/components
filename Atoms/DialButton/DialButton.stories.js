import React from 'react'

import { storiesOf } from '@storybook/react'
import centered from '@storybook/addon-centered/react'
import { withKnobs, select, boolean } from '@storybook/addon-knobs'

import DialButton from './DialButton'
import { USER_STATUS_AVAILABLE, USER_STATUS_BUSY, USER_STATUS_AWAY, USER_STATUS_OFFLINE } from '../../../constants'

const stories = storiesOf('Provider/v1/Atoms', module)
stories.addDecorator(withKnobs)

stories.addDecorator(centered).add('DialButton', () => {
  const statusLabel = 'Size'
  const statusOptions = {
    available: USER_STATUS_AVAILABLE,
    busy: USER_STATUS_BUSY,
    away: USER_STATUS_AWAY,
    offline: USER_STATUS_OFFLINE,
  }
  const statusDefaultValue = USER_STATUS_AVAILABLE
  const statusValue = select(statusLabel, statusOptions, statusDefaultValue)

  const dimmedLab = 'dimmed'
  const dimmedDef = false
  const dimmedVal = boolean(dimmedLab, dimmedDef)

  return (
    <DialButton
      onClick={() => {
        alert('clicked')
      }}
      dimmed={dimmedVal}
      status={statusValue}
    />
  )
})
