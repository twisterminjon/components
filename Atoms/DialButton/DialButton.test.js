import React from 'react'

import DialButton from './DialButton'
import { USER_STATUS_AVAILABLE } from '../../../constants'

describe('DialButton', () => {
  const mockFunction = jest.fn()

  // Note: DialButton does additional determination of status internally,
  // determined by user context
  it('matches the snapshot if does not have mic and cam', () => {
    const wrapper = window.shallow(<DialButton onClick={mockFunction} dimmed={false} status={USER_STATUS_AVAILABLE} />)
    expect(wrapper).toMatchSnapshot()
  })
})
