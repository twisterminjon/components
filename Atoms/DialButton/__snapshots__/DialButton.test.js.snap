// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialButton matches the snapshot if does not have mic and cam 1`] = `
<button
  aria-label="dial"
  className="dialbutton dialbutton--disabled"
  data-testid="button-start-call"
  disabled={true}
  onClick={[MockFunction]}
  style={Object {}}
>
  <IconAnswerCall
    className=""
    color="var(--statusbubble_offline)"
    style={
      Object {
        "transform": "scale(0.7)",
      }
    }
  />
</button>
`;
